### PR TITLE
updated version of Add complex text support #1682

### LINKED ISSUE #1089: Bug in rendering Indic fonts
ImageFont, ImageDraw do not render complex scripts with ligatures correctly. (On Ubuntu 14.04, Python 3.4.0)

This is how it looks in editor gedit:
gedit
This is how Pillow does it.
pillow

```python
#!/usr/bin/env python3
#-*- coding:utf8 -*-

from PIL import Image, ImageFont, ImageDraw

im = Image.new("RGB",(160, 160))
draw = ImageDraw.Draw(im)

font_telugu = ImageFont.truetype("/usr/share/fonts/truetype/ttf-indic-fonts-core/Pothana2000.ttf",50)
text_telugu = "నిత్య"

font_hindi = ImageFont.truetype("/usr/share/fonts/truetype/ttf-indic-fonts-core/gargi.ttf",50)
text_hindi = "नित्य"

draw.text((10, 10), text_telugu, font=font_telugu)
draw.text((10, 90), text_hindi, font=font_hindi)

im.show()
```

### PR DESCRIPTION
**Pull request by Fahad-Alsaidi**
This pull request adds support for languages that require complex text layout.

We are using the Raqm library, that wraps FriBidi (for bidirectional
text support) and HarfBuzz (for text shaping), and does proper BiDi and script
itemization:
https://github.com/HOST-Oman/libraqm

This should fix #1089, #2255

@andreymal

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -16,9 +16,13 @@ python:
   - 3.4
   - nightly
 
+sudo: required # needed for trusty beta
```
Comment (khaledhosny) [inline .travis.yml]: I don’t think this is needed anymore, see https://docs.travis-ci.com/user/trusty-ci-environment/.

**Comment by Fahad-Alsaidi** [comment]
Finally, passed all tests. I welcome any feedback.

File: _imagingft.c [hunk 2]
```diff
@@ -321,6 +341,206 @@ font_getabc(FontObject* self, PyObject* args)
...
+    FT_ULong ch;
+    Py_ssize_t count;
+    FT_GlyphSlot glyph;
+    FT_Bool kerning = FT_HAS_KERNING(self->face);
+    FT_UInt last_index = 0;
+    int i;
+
+    if (features != Py_None || dir != NULL)
+      PyErr_SetString(PyExc_KeyError, "setting text direction or font features is not supported without libraqm");
+
```
Comment (hugovk) [inline _imagingft.c]: Add brackets for the `if` even for single statements. There's less chance of a bug later if another statement is added but the brackets forgotten.

File: _imagingft.c [hunk 3]
```diff
@@ -321,6 +341,206 @@ font_getabc(FontObject* self, PyObject* args)
...
+#endif
+        PyErr_SetString(PyExc_TypeError, "expected string");
+        return 0;
+    }
+
+    count = 0;
+    while (font_getchar(string, count, &ch))
+       count++;
+    if (count == 0)
+        return 0;
```
Comment (hugovk) [inline _imagingft.c]: Add brackets for `if`.

File: _imagingft.c [hunk 4]
```diff
@@ -321,6 +341,206 @@ font_getabc(FontObject* self, PyObject* args)
...
+#else
+    if (!PyUnicode_Check(string) && !PyString_Check(string)) {
+#endif
+        PyErr_SetString(PyExc_TypeError, "expected string");
+        return 0;
+    }
+
+    count = 0;
+    while (font_getchar(string, count, &ch))
+       count++;
```
Comment (hugovk) [inline _imagingft.c]: Add brackets for `while`.

File: _imagingft.c [hunk 5]
```diff
@@ -321,6 +341,206 @@ font_getabc(FontObject* self, PyObject* args)
...
+
+    (*glyph_info) = PyMem_New(GlyphInfo, count);
+    if ((*glyph_info) == NULL) {
+        PyErr_SetString(PyExc_MemoryError, "PyMem_New() failed");
+        return 0;
+    }
+
+    load_flags = FT_LOAD_RENDER|FT_LOAD_NO_BITMAP;
+    if (mask)
+        load_flags |= FT_LOAD_TARGET_MONO;
```
Comment (hugovk) [inline _imagingft.c]: Add brackets for `if`.

File: _imagingft.c [hunk 6]
```diff
@@ -188,60 +205,69 @@ font_getchar(PyObject* string, int index, FT_ULong* char_out)
     return 0;
 }
 
+static size_t
+text_layout(PyObject* string, FontObject* self, const char* dir,
+            PyObject *features ,GlyphInfo **glyph_info, int mask);
```
Comment (hugovk) [inline _imagingft.c]: Space after comma, not before.

File: _imagingft.c [hunk 7]
```diff
@@ -329,29 +549,26 @@ font_render(FontObject* self, PyObject* args)
...
-#if PY_VERSION_HEX >= 0x03000000
-    if (!PyUnicode_Check(string)) {
-#else
-    if (!PyUnicode_Check(string) && !PyString_Check(string)) {
-#endif
-        PyErr_SetString(PyExc_TypeError, "expected string");
+    glyph_info = NULL;
+    count = text_layout(string, self, dir, features, &glyph_info, mask);
+    if (count == 0)
         return NULL;
```
Comment (hugovk) [inline _imagingft.c]: Add brackets for `if`.

File: _imagingft.c [hunk 8]
```diff
@@ -329,29 +549,26 @@ font_render(FontObject* self, PyObject* args)
...
     int temp;
     int xx, x0, x1;
-    if (!PyArg_ParseTuple(args, "On|i:render", &string, &id, &mask))
+    const char *dir = NULL;
+    size_t count;
+    GlyphInfo *glyph_info;
+    PyObject *features = NULL;
+
+    if (!PyArg_ParseTuple(args, "On|izO:render", &string,  &id, &mask, &dir, &features))
         return NULL;
```
Comment (hugovk) [inline _imagingft.c]: Add brackets for `if`.

**Comment by wiredfool** [comment]
I've done a quick review of this PR. In general, this covers most of the issues that I had with the initial PR. 

I'm still concerned with the dependency on the not well known library raqm, especially on platforms other than Ubuntu Trusty. 

* Documentation is still too sparse:
  - There are three new libraries, only one is mentioned. 
  - The parameter values for direction and features are unspecified. 
* `depends/*` needs to be touched up and tested where possible.
* I think that the `text_layout` function should be broken up into two implementations, `text_layout_raqm` and `text_layout_[something else]` and use a #ifdef in `text_layout` to delegate to one or the other. Furthermore, I think splitting out `text_layout_raqm` to a separate source file makes sense to minimize the implementation specific changes to _imagingft.
* There are a lot of places where you've got PY_VERSION_HEX #ifdefs. Check to see if the definitions in `py3.h` cover the use cases. If they don't, maybe we should look at that.

**Comment by khaledhosny** [comment]
> I'm still concerned with the dependency on the not well known library raqm, especially on platforms other than Ubuntu Trusty.

The alternative would be to roll your own solution (not essentially a bad thing, but more code for your to maintain) or depend on large package with even larger dependencies like Pango (again not essentially bad, if you can afford Pango then it is indeed the more mature and featureful choice).

File: .travis.yml [hunk 9]
```diff
@@ -7,7 +7,7 @@ notifications:
 # Run latest 3.x and 2.x next, to get quick compatibility results.
 # Then run the remainder.
 python:
-  - "pypy"
+  - "pypy-5.3.1"
```
Comment (Fahad-Alsaidi) [inline .travis.yml]: Since we switch to Trusty, the default version of pypy in Tursty is PyPy 5.4.1. There is a well-known error in it : `TypeError: PyPy does not yet implement the new buffer interface`  see this bug #2163.

Comment (hugovk) [inline .travis.yml]: @Fahad-Alsaidi That looks good (I've edited it slightly). Please could you add it as a comment in .travis.yml, above `- "pypy-5.3.1"`?

Comment (wiredfool) [inline .travis.yml]: If we've got one test that is failing on a specific version of pypy, we should document that at the test and skip it for just that version.  We've done it before, and recently.

Comment (Fahad-Alsaidi) [inline .travis.yml]: @wiredfool  but the test not related to adding CTL support. Is it fine to edit that test or you may  edit later?

**Comment by Fahad-Alsaidi** [comment]
@wiredfool 
> There are three new libraries, only one is mentioned.

what is the right place to mention them?

> The parameter values for direction and features are unspecified.

I hope now they are clear, if not please give me a guideline. 

>  depends/* needs to be touched up and tested where possible.

I update all files but need testing at least in FreeBSD & Fedora. BTW, Fedora have already a package for  libraqm.

> I think that the text_layout function should be broken up into two implementations, text_layout_raqm and text_layout_[something else] and use a #ifdef in text_layout to delegate to one or the other. Furthermore, I think splitting out text_layout_raqm to a separate source file makes sense to minimize the implementation specific changes to _imagingft.

I spite text_layout to two implementations as you request but I don't think it is needed to put it to a separate file.

> There are a lot of places where you've got PY_VERSION_HEX #ifdefs. Check to see if the definitions in py3.h cover the use cases. If they don't, maybe we should look at that.

I really need help in this point.

@hugovk I did all modifications as requested. Please review.

File: .travis.yml (hunk 9 above)
Comment (wiredfool) [inline .travis.yml]: That can be a separate PR, and note that it has to be merged prior to this one.

File: PIL/features.py [hunk 10]
```diff
@@ -5,6 +5,7 @@
     "tkinter": "PIL._imagingtk",
     "freetype2": "PIL._imagingft",
     "littlecms2": "PIL._imagingcms",
+    "raqm": "PIL._imagingft",
```
Comment (wiredfool) [inline PIL/features.py]: This is going to trigger on raqm being available when it's not compiled in but freetype is.

File: Tests/test_imagefontctl.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+from PIL import ImageDraw
+
+FONT_SIZE = 20
+FONT_PATH = "Tests/fonts/DejaVuSans.ttf"
+
+try:
+    from PIL import ImageFont
+
+    # check if raqm is available
+    ttf = ImageFont.truetype(FONT_PATH, FONT_SIZE)
```
Comment (wiredfool) [inline Tests/test_imagefontctl.py]: There should be a way to check if raqm is installed. As it stands, if there is an error that triggers an import error on ImageFont, then this whole set of tests are going to get skipped.

There are version flags for several of the libraries, that would be something to add to _imagingft.c, potentially for all three libraries, but at least raqm.

File: _imagingft.c [hunk 12]
```diff
@@ -41,6 +41,23 @@
...
+#include <raqm.h>
+#else
+typedef enum
+{
+  RAQM_DIRECTION_DEFAULT,
+  RAQM_DIRECTION_RTL,
+  RAQM_DIRECTION_LTR,
+  RAQM_DIRECTION_TTB
+} raqm_direction_t;
+#endif
```
Comment (wiredfool) [inline _imagingft.c]: It doesn't look like this is ever used outside of the HAVE_RAQM  #ifdef

File: _imagingft.c [hunk 13]
```diff
@@ -188,60 +205,284 @@ font_getchar(PyObject* string, int index, FT_ULong* char_out)
...
     FT_UInt last_index = 0;
+    int i;
 
-    /* calculate size and bearing for a given string */
-
-    PyObject* string;
-    if (!PyArg_ParseTuple(args, "O:getsize", &string))
-        return NULL;
-
+    if (features != Py_None || dir != NULL) {
```
Comment (wiredfool) [inline _imagingft.c]: This caveat needs to make it into the documentation

File: _imagingft.c [hunk 14]
```diff
@@ -188,60 +205,284 @@ font_getchar(PyObject* string, int index, FT_ULong* char_out)
...
+
+static size_t
+text_layout(PyObject* string, FontObject* self, const char* dir,
+            PyObject *features, GlyphInfo **glyph_info, int mask)
+{
+    size_t count;
+#ifdef HAVE_RAQM
+    count = text_layout_raqm(string, self, dir, features, glyph_info,  mask);
+#else
+    count = text_layout_fallback(string, self, dir, features, glyph_info, mask);
```
Comment (wiredfool) [inline _imagingft.c]: I know I said #ifdef earlier, but this makes testing the fallback harder. I'd really like to have automated tests on both of them...

File: depends/ubuntu_14.04.sh [hunk 15]
```diff
@@ -4,12 +4,14 @@
 # Installs all of the dependencies for Pillow for Ubuntu 14.04
 # for both system Pythons 2.7 and 3.4
 #
-
+sudo add-apt-repository -y ppa:as-bahanta/raqm
+sudo apt-get update
```
Comment (wiredfool) [inline depends/ubuntu_14.04.sh]: This is extraneous with the ./install_raqm line below.

File: depends/install_raqm.sh [hunk 16]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+# install raqm
+
+
+if [ ! -f raqm-0.2.0.tar.gz ]; then
+    wget -O 'raqm-0.2.0.tar.gz' 'https://github.com/HOST-Oman/libraqm/releases/download/v0.2.0/raqm-0.2.0.tar.gz?raw=true'
```
Comment (wiredfool) [inline depends/install_raqm.sh]: Should add a PR to add this to our depends repo.

File: docs/installation.rst [hunk 17]
```diff
@@ -170,6 +170,8 @@ Many of Pillow's features require external libraries:
   * Windows support: Libimagequant requires VS2013/MSVC 18 to compile,
     so it is unlikely to work with any Python prior to 3.5 on Windows.
 
+* **libraqm** provides complex text layout support.
+
```
Comment (wiredfool) [inline docs/installation.rst]: Need to mention the other libraries. Also, below there are build flags that need to be updated.

File: docs/reference/ImageDraw.rst [hunk 18]
```diff
@@ -240,9 +240,15 @@ Methods
                     the number of pixels between lines.
     :param align: If the text is passed on to multiline_text(),
                   "left", "center" or "right".
+    :param direction: Direction of the text. It can be  'rtl', 'ltr', 'ttb' or 'btt'.
```
Comment (wiredfool) [inline docs/reference/ImageDraw.rst]: Needs a Requires raqm note here, and all similar locations.

File: setup.py [hunk 19]
```diff
@@ -131,7 +131,7 @@ def _pkg_config(name):
 
 class pil_build_ext(build_ext):
     class feature:
-        features = ['zlib', 'jpeg', 'tiff', 'freetype', 'lcms', 'webp',
+        features = ['zlib', 'jpeg', 'tiff', 'freetype', 'raqm', 'lcms', 'webp',
```
Comment (wiredfool) [inline setup.py]: need to filter and error if raqm is requested and freetype is either not found or disabled.

**Comment by wiredfool** [comment]
Libraries should be mentioned in docs/installation.rst. It should be clear from the discussion there what libraries we rely on, as well as filling in the examples that are there. There are also the build flags that need to be updated. 

Do you know what the status of the dependencies are on Windows and OSX?

**Comment by wiredfool** [comment]
@hugovk We do need to mention the licence for the font somewhere.

File: PIL/ImageDraw.py [hunk 20]
```diff
@@ -217,30 +217,27 @@ def _multiline_split(self, text):
...
         return text.split(split_character)
 
-    def text(self, xy, text, fill=None, font=None, anchor=None,
-             *args, **kwargs):
+    def text(self, xy, text, fill=None, font=None, anchor=None, *args, **kwargs):
         if self._multiline_check(text):
-            return self.multiline_text(xy, text, fill, font, anchor,
-                                       *args, **kwargs)
-
+            return self.multiline_text(xy, text, fill, font, anchor, *args, **kwargs)
```
Comment (homm) [inline PIL/ImageDraw.py]: pep8

File: .travis.yml (hunk 9 above)
Comment (Fahad-Alsaidi) [inline .travis.yml]: @wiredfool  done

File: _imagingft.c (hunk 14 above)
Comment (Fahad-Alsaidi) [inline _imagingft.c]: Right now we have two test one for ctl which is test_imagefont.py and for CTL which is test_imagefontctl.py. Are they not enough ?

File: PIL/features.py (hunk 10 above)
Comment (Fahad-Alsaidi) [inline PIL/features.py]: when I delete it, all builds fails the error is 
```
Traceback (most recent call last):
  File "selftest.py", line 184, in <module>
    supported = features.check_module(name)
  File "build\bdist.win-amd64\egg\PIL\features.py", line 15, in check_module
ValueError: Unknown module raqm
```

File: _imagingft.c (hunk 14 above)
Comment (wiredfool) [inline _imagingft.c]: Say an existing user runs `pip install --upgrade pillow` without installing raqm. Is the new text rendering code they will be running tested?

File: depends/ubuntu_14.04.sh (hunk 15 above)
Comment (wiredfool) [inline depends/ubuntu_14.04.sh]: It doesn't work with the version in trusty?

Comment (Fahad-Alsaidi) [inline depends/ubuntu_14.04.sh]: fixed now.

File: depends/install_raqm.sh (hunk 16 above)
Comment (hugovk) [inline depends/install_raqm.sh]: PR created: https://github.com/python-pillow/pillow-depends/pull/17

File: .travis.yml (hunk 9 above)
Comment (hugovk) [inline .travis.yml]: Fixed via PR https://github.com/python-pillow/Pillow/pull/2294.

**Comment by wiredfool** [comment]
Closed, merged with #2576 

Thanks to everyone who worked on this, I think it's a valuable addition to the library.
